### PR TITLE
Add SPV_INTEL_masked_gather_scatter extension

### PR DESCRIPTION
**Pull request by MrSidims**
This extension allows TypeVector to have a Physical Pointer Type
Component Type and introduces gather/scatter instructions.
It will be useful for explicitly vectorized kernels.

Spec: https://github.com/intel/llvm/pull/6613

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

### CODE REVIEW

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 1]
```diff
@@ -3395,6 +3395,132 @@ class SPIRVComplexFloatInst
...
+    SPIRVType *ResTy = this->getType();
+    SPVErrLog.checkError(ResTy->isTypeVector(), SPIRVEC_InvalidInstruction,
+                         InstName + "\nResult must be a vector type\n");
+    SPIRVWord ResCompCount = ResTy->getVectorComponentCount();
+    SPIRVType *ResCompTy = ResTy->getVectorComponentType();
+
+    SPIRVValue *PtrVec =
+        const_cast<SPIRVMaskedGatherINTELInst *>(this)->getOperand(0);
+    SPIRVType *PtrVecTy = PtrVec->getType();
+    SPVErrLog.checkError(
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3424]: Oh, now I remember. I wanted to rewrite these checks and may be move them to SPIRVSIMDINTELInstBase

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 2]
```diff
@@ -3395,6 +3395,132 @@ class SPIRVComplexFloatInst
 _SPIRV_OP(ComplexFMulINTEL)
 _SPIRV_OP(ComplexFDivINTEL)
 #undef _SPIRV_OP
+
+class SPIRVSIMDINTELInstBase : public SPIRVInstTemplateBase {
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3399]: Probably should be renamed

**Comment by MrSidims** [comment]
@asudarsa @vmaksimo @jcranmer-intel please take a look

File: lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -1026,6 +1026,8 @@ Value *SPIRVToLLVM::transConvertInst(SPIRVValue *BV, Function *F,
   switch (BC->getOpCode()) {
   case OpPtrCastToGeneric:
   case OpGenericCastToPtr:
+    CO = Instruction::AddrSpaceCast;
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVReader.cpp:1029]: Why is this change required?

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -2122,7 +2124,12 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   case OpInBoundsPtrAccessChain: {
     auto AC = static_cast<SPIRVAccessChainBase *>(BV);
     auto Base = transValue(AC->getBase(), F, BB);
-    Type *BaseTy = transType(AC->getBase()->getType()->getPointerElementType());
+    SPIRVType *BaseSPVTy = AC->getBase()->getType();
+    Type *BaseTy =
+        BaseSPVTy->isTypeVector()
+            ? transType(
+                  BaseSPVTy->getVectorComponentType()->getPointerElementType())
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVReader.cpp:2131]: Can this logic be safely moved inside getPointerElementType?

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 5]
```diff
@@ -3395,6 +3395,132 @@ class SPIRVComplexFloatInst
 _SPIRV_OP(ComplexFMulINTEL)
 _SPIRV_OP(ComplexFDivINTEL)
 #undef _SPIRV_OP
+
+class SPIRVGatherScatterINTELInstBase : public SPIRVInstTemplateBase {
```
Comment (asudarsa) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3399]: Maybe SPIRVMaskedGatherScatterINTELInstBase is a more accurate name? (Just a nit)

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 6]
```diff
@@ -3395,6 +3395,132 @@ class SPIRVComplexFloatInst
...
+            PtrVecCompCount == FillEmptyCompCount &&
+            FillEmptyCompCount == MaskCompCount,
+        SPIRVEC_InvalidInstruction,
+        InstName + "\nResult, PtrVector, Mask and FillEmpty vectors must have "
+                   "the same size\n");
+
+    SPVErrLog.checkError(
+        ResCompTy == PtrElemTy && PtrElemTy == FillEmptyCompTy,
+        SPIRVEC_InvalidInstruction,
+        InstName + "\nResult, PtrVector and FillEmpty vectors must have "
```
Comment (asudarsa) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3462]: This needs to be: Component Type of Result and FillEmpty vector must be same as base type of PtrVector.

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp [hunk 7]
```diff
@@ -543,11 +543,17 @@ bool SPIRVRegularizeLLVMBase::regularize() {
...
-            PointerType *InterTy = PointerType::getWithSamePointeeType(
-                DestTy, SrcTy->getPointerAddressSpace());
+          Type *DestTy = ASCast->getDestTy();
+          Type *SrcTy = ASCast->getSrcTy();
+          if (!II.getContext().supportsTypedPointers())
+            continue;
+          if (DestTy->getScalarType()->getNonOpaquePointerElementType() !=
+              SrcTy->getScalarType()->getNonOpaquePointerElementType()) {
+            Type *InterTy = PointerType::getWithSamePointeeType(
+                cast<PointerType>(DestTy), SrcTy->getPointerAddressSpace());
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVRegularizeLLVM.cpp:553]: cast<PointerType> for SrcTy?

File: lib/SPIRV/SPIRVWriter.cpp [hunk 8]
```diff
@@ -3824,6 +3843,47 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
     }
     return Op;
   }
+  case Intrinsic::masked_gather: {
+    if (BM->isAllowedToUseExtension(
```
Comment (asudarsa) [inline lib/SPIRV/SPIRVWriter.cpp:3847]: Why do we have to check this again? Is there a possibility that we would not have called transType for FixedVectorType before this?

**Comment by asudarsa** [comment]
Added some comments. Also added a coupe of comments in the doc PR.

Thanks

File: lib/SPIRV/SPIRVWriter.cpp (hunk 8 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVWriter.cpp:3847]: That's unlikely, but it will explicitly say to the one, who is ready this code next, that these instructions are only supported as a part of the appropriate extension. For the sake of performance I can rewrite these line in the following way:
if (notSupported)
   error_out
   return
do_translation

File: lib/SPIRV/SPIRVReader.cpp (hunk 3 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVReader.cpp:1029]: It's an artifact, removed it.

File: lib/SPIRV/SPIRVReader.cpp (hunk 4 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVReader.cpp:2131]: It's a very good suggestion. But after some considering I'd say: no.

The reasoning, that it's safe to move it now, but I can imagine, that in the future some vendor, which is not aware or interested in our extension, can add another instructions, that are operating over pointers. The change you are suggesting would implicitly allow vector of pointers to be used by those instruction, which is incorrect.

And may be it's already the case.

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp (hunk 7 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVRegularizeLLVM.cpp:553]: Note sure if I get the question/suggestion.

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h (hunk 6 above)
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h]: Thanks, applied

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h (hunk 5 above)
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h]: Applied

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 9]
```diff
@@ -3395,6 +3395,132 @@ class SPIRVComplexFloatInst
...
+        InstName + "\nPtrVector must be a vector of pointers type\n");
+    SPIRVWord PtrVecCompCount = PtrVecTy->getVectorComponentCount();
+    SPIRVType *PtrVecCompTy = PtrVecTy->getVectorComponentType();
+    SPIRVType *PtrElemTy = PtrVecCompTy->getPointerElementType();
+
+    SPVErrLog.checkError(
+        this->isOperandLiteral(1), SPIRVEC_InvalidInstruction,
+        InstName + "\nAlignment must be a constant expression integer\n");
+
+    SPIRVValue *Mask =
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:3460]: Added the check

File: test/transcoding/SPV_INTEL_masked_gather_scatter/intel-basic-vector-pointers.ll [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+entry:
+  %arg1 = alloca <4 x i8 addrspace(4)*>
+  %arg2 = alloca <4 x i8 addrspace(4)*>
+  %0 = load <4 x i8 addrspace(4)*>, <4 x i8 addrspace(4)*>* %arg1
+  store <4 x i8 addrspace(4)*> %0, <4 x i8 addrspace(4)*>* %arg2
+  %tmp1 = bitcast <4 x i8 addrspace(4)*> %0 to <4 x i32 addrspace(4)*>
+  %tmp2 = addrspacecast <4 x i32 addrspace(4)*> %tmp1 to  <4 x i32 addrspace(1)*>
+  %tmp3 = call <4 x i32 addrspace(1)*> @boo(<4 x i32 addrspace(1)*> %tmp2)
+  %tmp4 = getelementptr inbounds i32, <4 x i32 addrspace(1)*> %tmp3, i32 1
+  ret void
```
Comment (jcranmer-intel) [inline test/transcoding/SPV_INTEL_masked_gather_scatter/intel-basic-vector-pointers.ll:60]: Could you add an addrspacecast line that also bitcasts to this test, like this?
` %tmp5 = addrspacecast <4 x i32 addrspace(4)*> %tmp1 to <4 x i8 addrspace(1)*>`

Comment (MrSidims) [reply test/transcoding/SPV_INTEL_masked_gather_scatter/intel-basic-vector-pointers.ll:60]: Thanks! It actually reveals a bug in regularize pass.

File: lib/SPIRV/SPIRVRegularizeLLVM.cpp (hunk 7 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVRegularizeLLVM.cpp:553]: Got it. Added.

**Review by jcranmer-intel — APPROVED** [review]
I haven't looked too closely at the code in lib/SPIRV/libSPIRV, but otherwise, this looks good.

File: lib/SPIRV/SPIRVWriter.cpp (hunk 8 above)
Comment (asudarsa) [reply lib/SPIRV/SPIRVWriter.cpp]: I agree with the argument that it makes the code readable.

**Review by asudarsa — APPROVED** [review]
Thanks for addressing the comments. LGTM
